csrc/cpu/neighbor_sample_cpu.cpp

```diff
@@ -199,31 +224,58 @@ hetero_sample(const vector<node_t> &node_types,
           continue;
 
         if ((num_samples < 0) || (!replace && (num_samples >= col_count))) {
+          // select all neighbors
           for (int64_t offset = col_start; offset < col_end; offset++) {
             const int64_t &v = row_data[offset];
+            bool time_constraint = true; // whether src -> dst obeys the time constraint
+            if (is_temporal) {
+              const auto *src_time = node_time_dict.at(src_node_type).data_ptr<int64_t>();
+              if (dst_time < src_time[v])
+                time_constraint = false;
+            }
+            if (!time_constraint)
+              continue;
             const auto res = to_local_src_node.insert({v, src_samples.size()});
-            if (res.second)
+            if (res.second) {
               src_samples.push_back(v);
+              if (is_temporal)
+                src_root_time.push_back(dst_time);
+            }
             if (directed) {
               cols.push_back(i);
               rows.push_back(res.first->second);
               edges.push_back(offset);
             }
           }
         } else if (replace) {
-          for (int64_t j = 0; j < num_samples; j++) {
+          // sample with replacement
+          int64_t num_neighbors = 0;
+          while (num_neighbors < num_samples) {
             const int64_t offset = col_start + rand() % col_count;
             const int64_t &v = row_data[offset];
+            bool time_constraint = true; // whether src -> dst obeys the time constraint
+            if (is_temporal) {
+              const auto *src_time = node_time_dict.at(src_node_type).data_ptr<int64_t>();
```

Comment:
This one is the same as in the previous `if` and next `else`, so it could be placed after we get `src_node_type`

Reply:
thx! I did the refactor